hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java

```diff
@@ -147,10 +152,45 @@ private void initializePipelineState() throws IOException {
     }
   }
 
+  private boolean exceedPipelineNumberLimit(ReplicationFactor factor) {
+    if (factor != ReplicationFactor.THREE) {
+      // Only put limits for Factor THREE pipelines.
+      return false;
+    }
+    // Per datanode limit
+    if (heavyNodeCriteria > 0) {
+      return (stateManager.getPipelines(ReplicationType.RATIS, factor).size() -
```

Comment:
Or should we consider only OPEN pipeline here?

Reply:
Other pipelines like STALE pipelines would possibly turn into OPEN at any time. So here I just deduct out CLOSED pipelines.